MultiSource/Benchmarks/Ptrdist/anagram/anagram.c

```diff
@@ -223,24 +226,32 @@ typedef struct {
 } Letter;
 typedef Letter * PLetter;
 
-Letter alPhrase[ALPHABET]; /* statistics on the current phrase */
+// CHECKEDC : checkedc array type
+// checkedc array type of T == array pointer type of T
+// bounds information for checkedc array is from array size
+Letter alPhrase _Checked [ALPHABET]; /* statistics on the current phrase */
 #define lPhrase(ch) alPhrase[ch2i(ch)]  /* quick access to a letter */
 
 int cchPhraseLength;                    /* number of letters in phrase */
 
-Quad aqMainMask[MAX_QUADS];/* the bit field for the full phrase */
-Quad aqMainSign[MAX_QUADS];/* where the sign bits are */
+Quad aqMainMask _Checked [MAX_QUADS];/* the bit field for the full phrase */
+Quad aqMainSign _Checked [MAX_QUADS];/* where the sign bits are */
+
+char achPhrase _Checked [255];
 
 int cchMinLength = 3;
 
 /* auGlobalFrequency counts the number of times each letter appears, summed
  * over all candidate words.  This is used to decide which letter to attack
  * first.
  */
-unsigned auGlobalFrequency[ALPHABET];
-char achByFrequency[ALPHABET];          /* for sorting */
+unsigned auGlobalFrequency _Checked [ALPHABET];
+char achByFrequency _Checked [ALPHABET];          /* for sorting */
 
-char * pchDictionary;               /* the dictionary is read here */
+// CHECKEDC : externally scoped variable bounds information
+// global variable bounds information is defined in declaration
+// dictionary maximum length is defined as MAXWORDS
+_Array_ptr<char> pchDictionary : count(MAXWORDS);               /* the dictionary is read here */
```

Comment:
The bounds declaration for a global variable is a program invariant, similar to a Hoare-style program invariant.   The values of the variables used in the bounds declaration can change, so long as the bounds declaration is valid after changing the variables.  This means that you can update global variables with bounds declaration, so long as the bounds declaration is valid at the end of the updates.

For now, you can use the code that I suggested.
```
pchDictionarySize = ...
pchDictionary = malloc(pchDictionarySize)
```

The full answer is more complicated  We have not implemented the features in the compiler that support the full answer.    There is a question about when bounds declarations must be valid.  Normally, bounds declarations must be valid at the end of every statement.    A programmer may need several statements to update the global variables involved in a bounds declarations, so this would be problem.

For example, in
```
pchDictionarySize = ..... 
pchDictionary = malloc(pchDictionarySize)
```
there would be a problem after the update to `pchDictionarySize` because the bounds declaration is temporarily not valid.

We have a solution for this problem; `bundled` blocks.   In a `bundled block`, bounds declarations must be valid only at the end of the bundled block.    Thus, updates to global variable that has a bounds declarations and the global variables that are used in the bounds declaration can be included in a bundled block
```
bundle {
    pchDictonarySize = ...
    pch = malloc(...)
}
```
We have not implemented bundled blocks yet.